python/lsst/pipe/base/quantum_provenance_graph.py

```diff
@@ -1195,245 +1286,562 @@ def get_dataset_info(self, key: DatasetKey) -> DatasetInfo:
+                    f"Status went from successful in run {list(quantum_info['runs'].values())[-1]!r} "
+                    f"to {quantum_run.status!r} in {output_run!r}."
+                )
+            # If a quantum status is unknown and it moves to blocked, we
+            # know for sure that it is a blocked quantum.
+            case (QuantumInfoStatus.UNKNOWN, QuantumRunStatus.BLOCKED):
+                new_status = QuantumInfoStatus.BLOCKED
+            # A transition into blocked does not change the overall quantum
+            # status for a failure.
+            case (_, QuantumRunStatus.BLOCKED):
+                new_status = last_status
+            # If a quantum transitions from any state into missing
+            # metadata, we don't have enough information to diagnose its
+            # state.
+            case (_, QuantumRunStatus.METADATA_MISSING):
+                new_status = QuantumInfoStatus.UNKNOWN
+            # Any transition into failure is a failed state.
+            case (_, QuantumRunStatus.FAILED):
+                new_status = QuantumInfoStatus.FAILED
+        # Update `QuantumInfo.status` for this quantum.
+        quantum_info["status"] = new_status
+        return new_status
+
+    def _update_caveats(
+        self,
+        quantum_key: QuantumKey,
+        output_run: str,
+        read_caveats: Literal["lazy", "exhaustive"],
+        executor: concurrent.futures.Executor,
+        futures: list[concurrent.futures.Future[None]],
+    ) -> None:
+        """Read quantum success caveats and exception information from task
+        metadata.
+
+        Parameters
+        ----------
+        quantum_key : `QuantumKey`
+            Key for the node in the provenance graph.
+        output_run : `str`
+            Output run collection.
+        read_caveats : `str`
+            Whether to read metadata files to get flags that describe qualified
+            successes. If "exhaustive", all metadata files will be read.  If
+            "lazy", only metadata files where at least one predicted output is
+            missing will be read.
+        butler : `lsst.daf.butler.Butler`
+            The Butler used for this report. This should match the Butler
+            used for the run associated with the executed quantum graph.
+        """
+        if read_caveats == "lazy" and all(
+            self.get_dataset_info(dataset_key)["runs"][output_run].produced
+            for dataset_key in self._xgraph.successors(quantum_key)
+        ):
+            return
+        quantum_info = self.get_quantum_info(quantum_key)
+        quantum_run = quantum_info["runs"][output_run]
+
+        def read_metadata() -> None:
+            md = self._butler_get(quantum_run.metadata_ref, storageClass="TaskMetadata")
+            try:
+                # Int conversion guards against spurious conversion to
+                # float that can apparently sometimes happen in
+                # TaskMetadata.
```

Comment:
`TaskMetadata` is in some sense `dict[str, Any]` but I think the order of StrictInt vs StrictFloat matters.

https://github.com/lsst/pipe_base/blob/main/python/lsst/pipe/base/_task_metadata.py#L130

Reply:
I was wondering how this could happen.   Should I go reorder those on this ticket, and see if that lets me remove this?  I'm a little worried that could break something else (including things that might not be tested).

Reply:
It looks like the code has been `StrictFloat, StrictInt` from the beginning. Are you saying that `5` always comes back as `5.` because of that order?  I wonder if `5.` would be read as `5` if we flip the order. I think we used Strict versions because there were issues with numpy floats/ints in the serialization.

Reply:
I can't say for sure that it always happens, but I did observe it happen at least sometimes, and my only theory for what's going on would suggest that it always happens.

And I do think you're right about what would happen if we flipped it, which makes me think we should not flip it, since it's just trading one problem for another + potential disruption.